src/Data/Widget/Adapters/Adapter.elm

```diff
@@ -1,11 +1,48 @@
-module Data.Widget.Adapters.Adapter exposing (Adapter(..))
+module Data.Widget.Adapters.Adapter exposing (Adapter(..), decoder)
 
--- TODO these are hardcoded for now but will likely have to be a set of
--- defaults plus optional user defined adapters
--- e.g. "use row 3 for headings & row 4-6 for data"
+import Data.Widget.Adapters.MetricAdapter as MetricAdapter exposing (Config, defaultConfig, adapterConfigDecoder)
+import Json.Decode as Decode exposing (Decoder, maybe)
+import Json.Decode.Pipeline as Pipeline exposing (decode, required, optional)
+
+
+type alias Definition =
+    { type_ : String
+    , config : Maybe Config
+    }
 
 
 type Adapter
     = TABLE
     | BAR_CHART
     | HEAT_MAP
+    | METRIC MetricAdapter.Config
+
+
+decoder : Decoder Adapter
+decoder =
+    definitionDecoder
+        |> Decode.andThen
+            (\definition ->
+                case definition.type_ of
+                    "TABLE" ->
+                        Decode.succeed TABLE
+
+                    "BAR_CHART" ->
+                        Decode.succeed BAR_CHART
+
+                    "HEAT_MAP" ->
+                        Decode.succeed HEAT_MAP
+
+                    "METRIC" ->
+                        Decode.succeed <| METRIC (definition.config |> Maybe.withDefault MetricAdapter.defaultConfig)
+
+                    somethingElse ->
+                        Decode.fail <| "Unknown adapter: " ++ somethingElse
+            )
+
+
+definitionDecoder : Decoder Definition
+definitionDecoder =
+    decode Definition
+        |> required "type_" Decode.string
+        |> optional "config" (maybe adapterConfigDecoder) Nothing
```

Comment:
@msp The signature `adapterConfigDecoder` is `Decoder MetricAdapterConfig` so when we add config for other adapters this won't work. 

` |> optional "config" (maybe (oneOf [ adapterConfigDecoder, myOtherAdaptersConfigDecoder ])) Nothing`

☝️ This would work, but instead it seems like the decoder should be determined by the `type_` string or `Adapter` but I can't work out how to do that.

Reply:
Yep @robwebdev - good catch. I'm sure there's some decoder hoops we could jump through to figure that out but I wonder if we should simply define `Config` generically as a `Dict` (key, val pairs) and push the work down into the specific Adapter to ensure the correct keys exist? Then we can generally decode all `Config` with "Json.Decode.dict":http://package.elm-lang.org/packages/elm-lang/core/5.1.1/Json-Decode#dict  

I'm not sure if that's a bit too dynamic for Elm but it'd do the job?